xmake/rules/protobuf/proto.lua

```diff
@@ -169,37 +205,123 @@ function buildcmd(target, batchcmds, sourcefile_proto, opt, sourcekind)
-    local proto_rootdir
-    if #sourcebatch.sourcefiles > 0 then
-        local sourcefile = sourcebatch.sourcefiles[1]
-        local fileconfig = target:fileconfig(sourcefile)
+    -- get protoc
+    local protoc = _get_protoc(target, sourcekind)
+    
+    local sourcefiles = sourcebatch.sourcefiles
+    for _, sourcefile_proto in ipairs(sourcefiles) do
+        local dependfile = target:dependfile(sourcefile_proto)
+        depend.on_changed(function()
+            -- get c/c++ source file for protobuf
+            local prefixdir
+            local autogendir
+            local public
+            local grpc_cpp_plugin
+            local fileconfig = target:fileconfig(sourcefile_proto)
+            if fileconfig then
+                public = fileconfig.proto_public
+                prefixdir = fileconfig.proto_rootdir
+                -- custom autogen directory to access the generated header files
+                -- @see https://github.com/xmake-io/xmake/issues/3678
+                autogendir = fileconfig.proto_autogendir
+                grpc_cpp_plugin = fileconfig.proto_grpc_cpp_plugin
+            end
+            local rootdir = autogendir and autogendir or path.join(target:autogendir(), "rules", "protobuf")
+            local filename = path.basename(sourcefile_proto) .. ".pb" .. (sourcekind == "cxx" and ".cc" or "-c.c")
+            local sourcefile_cx = target:autogenfile(sourcefile_proto, {rootdir = rootdir, filename = filename})
+            local sourcefile_dir = prefixdir and path.join(rootdir, prefixdir) or path.directory(sourcefile_cx)
+        
+            local grpc_cpp_plugin_bin
+            local filename_grpc
+            local sourcefile_cx_grpc
+            if grpc_cpp_plugin then
+                grpc_cpp_plugin_bin = _get_grpc_cpp_plugin(target, sourcekind)
+                filename_grpc = path.basename(sourcefile_proto) .. ".grpc.pb.cc"
+                sourcefile_cx_grpc = target:autogenfile(sourcefile_proto, {rootdir = rootdir, filename = filename_grpc})
+            end
+        
+            local protoc_args = {
+                path(sourcefile_proto),
+                path(prefixdir and prefixdir or path.directory(sourcefile_proto), function (p) return "-I" .. p end),
+                path(sourcefile_dir, function (p) return (sourcekind == "cxx" and "--cpp_out=" or "--c_out=") .. p end)
+            }
+        
+            if grpc_cpp_plugin then
+                local extension = target:is_plat("windows") and ".exe" or ""
+                table.insert(protoc_args, "--plugin=protoc-gen-grpc=" .. grpc_cpp_plugin_bin .. extension)
+                table.insert(protoc_args, path(sourcefile_dir, function (p) return ("--grpc_out=") .. p end))
+            end
+
+            os.mkdir(sourcefile_dir)
+            if opt.progress then
+                progress.show(opt.progress, "${color.build.object}compiling.proto %s", sourcefile_proto)
+            end
+            os.vrunv(protoc, protoc_args)
```

Comment:
This is not correct, `before_build_files(), {batch = true}` is only used to generate parallel compiled task DAG graphs to batchjobs, which contain the full task dependencies. This script does not go and perform any compilation actions directly.

We should execute it in each batchjob, so we generate full batchjobs to execute them via module_parser.

Reply:
If you execute run directly here, then they will all be executed early and completely serially, and cannot be executed in parallel.